api/src/db/programs.rs

```diff
@@ -36,72 +50,168 @@ impl DbClient {
         use crate::schema::verified_programs::dsl::*;
-
         // Ensure page is valid
-        let page = if page > 0 { page } else { 1 };
+        let page = page.max(1);
         let offset = (page - 1) * PER_PAGE;
 
         let conn = &mut self.get_db_conn().await?;
 
-        verified_programs
+        let all_program_ids = verified_programs
+            .filter(is_verified.eq(true))
             .select(program_id)
-            .order_by(id)
-            .limit(PER_PAGE)
-            .offset(offset)
+            .distinct()
+            .order_by(program_id)
             .load::<String>(conn)
-            .await
-            .map_err(|e| {
-                error!("Failed to fetch verified programs: {}", e);
-                e.into()
+            .await?;
+
+        let client = Arc::new(RpcClient::new(CONFIG.rpc_url.clone()));
+
+        // Return only programs where the verification PDA signer is the program upgrade authority, with caching for validation
+        let valid_programs: Vec<String> = stream::iter(all_program_ids)
+            .map(|pid| {
+                let client = Arc::clone(&client);
+                async move {
+                    let cache_key = format!("valid_program:{}", pid);
+
+                    // Try cache first
+                    match self.get_cache(&cache_key).await {
+                        Ok(cached) if cached == "1" => return Some(pid),
+                        Ok(_) => return None,
+                        Err(_) => {}
+                    }
+
+                    // Check DB for saved authority
+                    let saved_authority = self
+                        .get_program_authority_from_db(&pid)
+                        .await
+                        .ok()
+                        .flatten();
+
+                    // Fallback to RPC if not in DB
+                    let authority = match saved_authority {
+                        Some(a) => Some(a),
+                        None => get_program_authority(&pid)
+                            .await
+                            .ok()
+                            .and_then(|(auth, _)| auth),
+                    };
+
+                    if let Some(program_authority) = authority {
+                        if let (Ok(program_pubkey), Ok(authority_pubkey)) = (
+                            Pubkey::try_from(pid.as_str()),
+                            Pubkey::try_from(program_authority.as_str()),
+                        ) {
+                            if get_otter_pda(&client, &authority_pubkey, &program_pubkey)
+                                .await
+                                .is_ok()
+                            {
+                                _ = self.set_cache(&cache_key, "1").await;
+                                return Some(pid);
+                            }
+                        }
+                    }
+
+                    _ = self.set_cache(&cache_key, "0").await;
```

Comment:
does this work for frozen programs?

Reply:
Looks like it does: `execXUrAsMnqMmTHj5m7N1YQgsDz3cwGLYCYyuDRciV ` in `/verified-programs/3`